File: src/test/java/no/nav/bidrag/beregn/samvaersfradrag/SamvaersfradragPeriodeTest.java
package no.nav.bidrag.beregn.samvaersfradrag;

import static no.nav.bidrag.beregn.TestUtil.SAMVAERSFRADRAG_REFERANSE;
import static org.assertj.core.api.Assertions.assertThat;
import static org.junit.jupiter.api.Assertions.assertAll;

import java.math.BigDecimal;
import java.time.LocalDate;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Comparator;
import java.util.List;
import no.nav.bidrag.beregn.felles.bo.Avvik;
import no.nav.bidrag.beregn.felles.bo.Periode;
import no.nav.bidrag.beregn.felles.bo.Sjablon;
import no.nav.bidrag.beregn.felles.bo.SjablonInnhold;
import no.nav.bidrag.beregn.felles.bo.SjablonNokkel;
import no.nav.bidrag.beregn.felles.bo.SjablonPeriode;
import no.nav.bidrag.beregn.felles.enums.AvvikType;
import no.nav.bidrag.beregn.felles.enums.SjablonInnholdNavn;
import no.nav.bidrag.beregn.felles.enums.SjablonNavn;
import no.nav.bidrag.beregn.felles.enums.SjablonNokkelNavn;
import no.nav.bidrag.beregn.samvaersfradrag.bo.BeregnSamvaersfradragGrunnlag;
import no.nav.bidrag.beregn.samvaersfradrag.bo.BeregnSamvaersfradragResultat;
import no.nav.bidrag.beregn.samvaersfradrag.bo.SamvaersfradragGrunnlagPeriode;
import no.nav.bidrag.beregn.samvaersfradrag.periode.SamvaersfradragPeriode;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;

class SamvaersfradragPeriodeTest {

  private final SamvaersfradragPeriode samvaersfradragPeriode = SamvaersfradragPeriode.getInstance();

  @Test
  @DisplayName("Test av periodisering. Resultatperioden skal være lik beregnDatoFra -> beregnDatoTil")
  void testPeriodisering() {
    var beregnDatoFra = LocalDate.parse("2020-06-01");
    var beregnDatoTil = LocalDate.parse("2020-07-01");

    // Lag samværsinfo
    var samvaersfradragGrunnlagPeriodeListe = new ArrayList<SamvaersfradragGrunnlagPeriode>();
    samvaersfradragGrunnlagPeriodeListe.add(new SamvaersfradragGrunnlagPeriode(SAMVAERSFRADRAG_REFERANSE,
        new Periode(LocalDate.parse("2019-01-01"), LocalDate.parse("2020-10-01")), 1, LocalDate.parse("2016-03-17"), "02"));

    samvaersfradragGrunnlagPeriodeListe.add(new SamvaersfradragGrunnlagPeriode(SAMVAERSFRADRAG_REFERANSE,
        new Periode(LocalDate.parse("2019-01-01"), LocalDate.parse("2020-10-01")), 2, LocalDate.parse("2017-05-17"), "02"));

    // Lag sjabloner
    var sjablonPeriodeListe = new ArrayList<SjablonPeriode>();

    sjablonPeriodeListe.add(new SjablonPeriode(
        new Periode(LocalDate.parse("2018-07-01"), LocalDate.parse("2020-06-30")),
        new Sjablon(SjablonNavn.SAMVAERSFRADRAG.getNavn(),
            Arrays.asList(new SjablonNokkel(SjablonNokkelNavn.SAMVAERSKLASSE.getNavn(), "02"),
                new SjablonNokkel(SjablonNokkelNavn.ALDER_TOM.getNavn(), "5")),
            Arrays.asList(new SjablonInnhold(SjablonInnholdNavn.ANTALL_DAGER_TOM.getNavn(), BigDecimal.ZERO),
                new SjablonInnhold(SjablonInnholdNavn.ANTALL_NETTER_TOM.getNavn(), BigDecimal.valueOf(8)),
                new SjablonInnhold(SjablonInnholdNavn.FRADRAG_BELOP.getNavn(), BigDecimal.valueOf(727))))));

    BeregnSamvaersfradragGrunnlag beregnSamvaersfradragGrunnlag = new BeregnSamvaersfradragGrunnlag(beregnDatoFra, beregnDatoTil,
        samvaersfradragGrunnlagPeriodeListe, sjablonPeriodeListe);

    var resultat = samvaersfradragPeriode.beregnPerioder(beregnSamvaersfradragGrunnlag);

    assertAll(
        () -> assertThat(resultat.getResultatPeriodeListe()).hasSize(1),
        () -> assertThat(resultat.getResultatPeriodeListe().get(0).getResultatBeregningListe()).hasSize(2),
        () -> assertThat(resultat.getResultatPeriodeListe().get(0).getResultatDatoFraTil().getDatoFom()).isEqualTo(LocalDate.parse("2020-06-01")),
        () -> assertThat(resultat.getResultatPeriodeListe().get(0).getResultatDatoFraTil().getDatoTil()).isEqualTo(LocalDate.parse("2020-07-01")),
        () -> assertThat(
            resultat.getResultatPeriodeListe().get(0).getResultatBeregningListe().get(0).getResultatSamvaersfradragBelop())
            .isEqualByComparingTo(BigDecimal.valueOf(727)),
        () -> assertThat(
            resultat.getResultatPeriodeListe().get(0).getResultatBeregningListe().get(1).getResultatSamvaersfradragBelop())
            .isEqualByComparingTo(BigDecimal.valueOf(727))
    );

    printGrunnlagResultat(resultat);
  }

  @Test
  @DisplayName("Test med feil i grunnlag som skal resultere i avvik")
  void testGrunnlagMedAvvik() {
    var beregnDatoFra = LocalDate.parse("2018-07-01");
    var beregnDatoTil = LocalDate.parse("2021-01-01");
    var barnPersonId = 1;
    var barnFodselsdato = LocalDate.parse("2014-03-17");

    // Lag samværsinfo
    var samvaersklassePeriodeListe = new ArrayList<SamvaersfradragGrunnlagPeriode>();
    samvaersklassePeriodeListe.add(new SamvaersfradragGrunnlagPeriode(SAMVAERSFRADRAG_REFERANSE,
        new Periode(LocalDate.parse("2019-01-01"), LocalDate.parse("2020-07-01")), barnPersonId, barnFodselsdato, "02"));

    // Lag sjabloner
    var sjablonPeriodeListe = new ArrayList<SjablonPeriode>();

    sjablonPeriodeListe.add(new SjablonPeriode(
        new Periode(LocalDate.parse("2018-07-01"), LocalDate.parse("2020-06-30")),
        new Sjablon(SjablonNavn.SAMVAERSFRADRAG.getNavn(),
            Arrays.asList(new SjablonNokkel(SjablonNokkelNavn.SAMVAERSKLASSE.getNavn(), "02"),
                new SjablonNokkel(SjablonNokkelNavn.ALDER_TOM.getNavn(), "5")),
            Arrays.asList(new SjablonInnhold(SjablonInnholdNavn.ANTALL_DAGER_TOM.getNavn(), BigDecimal.ZERO),
                new SjablonInnhold(SjablonInnholdNavn.ANTALL_NETTER_TOM.getNavn(), BigDecimal.valueOf(8)),
                new SjablonInnhold(SjablonInnholdNavn.FRADRAG_BELOP.getNavn(), BigDecimal.valueOf(727))))));
    sjablonPeriodeListe.add(new SjablonPeriode(
        new Periode(LocalDate.parse("2018-07-01"), null),
        new Sjablon(SjablonNavn.SAMVAERSFRADRAG.getNavn(),
            Arrays.asList(new SjablonNokkel(SjablonNokkelNavn.SAMVAERSKLASSE.getNavn(), "02"),
                new SjablonNokkel(SjablonNokkelNavn.ALDER_TOM.getNavn(), "10")),
            Arrays.asList(new SjablonInnhold(SjablonInnholdNavn.ANTALL_DAGER_TOM.getNavn(), BigDecimal.ZERO),
                new SjablonInnhold(SjablonInnholdNavn.ANTALL_NETTER_TOM.getNavn(), BigDecimal.valueOf(8)),
                new SjablonInnhold(SjablonInnholdNavn.FRADRAG_BELOP.getNavn(), BigDecimal.valueOf(1052))))));

    BeregnSamvaersfradragGrunnlag beregnSamvaersfradragGrunnlag = new BeregnSamvaersfradragGrunnlag(beregnDatoFra, beregnDatoTil,
        samvaersklassePeriodeListe, sjablonPeriodeListe);

    var avvikListe = samvaersfradragPeriode.validerInput(beregnSamvaersfradragGrunnlag);

    assertAll(
        () -> assertThat(avvikListe).isNotEmpty(),
        () -> assertThat(avvikListe).hasSize(2),

        () -> assertThat(avvikListe.get(0).getAvvikTekst())
            .isEqualTo("Første dato i samvaersklassePeriodeListe (2019-01-01) er etter beregnDatoFra (2018-07-01)"),
        () -> assertThat(avvikListe.get(0).getAvvikType()).isEqualTo(AvvikType.PERIODE_MANGLER_DATA),

        () -> assertThat(avvikListe.get(1).getAvvikTekst())
            .isEqualTo("Siste dato i samvaersklassePeriodeListe (2020-07-01) er før beregnDatoTil (2021-01-01)"),
        () -> assertThat(avvikListe.get(1).getAvvikType()).isEqualTo(AvvikType.PERIODE_MANGLER_DATA)
    );

    printAvvikListe(avvikListe);
  }


  private void printGrunnlagResultat(BeregnSamvaersfradragResultat beregnSamvaersfradragResultat) {
    beregnSamvaersfradragResultat.getResultatPeriodeListe().stream()
        .sorted(Comparator.comparing(pR -> pR.getResultatDatoFraTil().getDatoFom()))
        .forEach(sortedPR -> System.out
            .println("Dato fra: " + sortedPR.getResultatDatoFraTil().getDatoFom() + "; " + "Dato til: "
                + sortedPR.getResultatDatoFraTil().getDatoTil() + "; " + "Samvaersfradragsbeløp: "
                + sortedPR.getResultatBeregningListe().get(0).getResultatSamvaersfradragBelop()));
  }

  private void printAvvikListe(List<Avvik> avvikListe) {
    avvikListe.forEach(avvik -> System.out.println("Avvik tekst: " + avvik.getAvvikTekst() + "; " + "Avvik type: " + avvik.getAvvikType()));
  }
}

File: src/test/kotlin/no/nav/bidrag/beregn/samvaersfradrag/SamvaersfradragPeriodeTest.kt
package no.nav.bidrag.beregn.samvaersfradrag

import no.nav.bidrag.beregn.TestUtil
import no.nav.bidrag.beregn.felles.bo.Periode
import no.nav.bidrag.beregn.felles.bo.Sjablon
import no.nav.bidrag.beregn.felles.bo.SjablonInnhold
import no.nav.bidrag.beregn.felles.bo.SjablonNokkel
import no.nav.bidrag.beregn.felles.bo.SjablonPeriode
import no.nav.bidrag.beregn.samvaersfradrag.bo.BeregnSamvaersfradragGrunnlag
import no.nav.bidrag.beregn.samvaersfradrag.bo.SamvaersfradragGrunnlagPeriode
import no.nav.bidrag.beregn.samvaersfradrag.periode.SamvaersfradragPeriode.Companion.getInstance
import no.nav.bidrag.domene.enums.beregning.Avvikstype
import no.nav.bidrag.domene.enums.sjablon.SjablonInnholdNavn
import no.nav.bidrag.domene.enums.sjablon.SjablonNavn
import no.nav.bidrag.domene.enums.sjablon.SjablonNøkkelNavn
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Assertions.assertAll
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.function.Executable
import java.math.BigDecimal
import java.time.LocalDate

internal class SamvaersfradragPeriodeTest {

    private val samvaersfradragPeriode = getInstance()

    @Test
    @DisplayName("Test av periodisering. Resultatperioden skal være lik beregnDatoFra -> beregnDatoTil")
    fun testPeriodisering() {
        // Lag samværsinfo
        val samvaersfradragGrunnlagPeriodeListe = mutableListOf<SamvaersfradragGrunnlagPeriode>()
        samvaersfradragGrunnlagPeriodeListe.add(
            SamvaersfradragGrunnlagPeriode(
                referanse = TestUtil.SAMVÆRSFRADRAG_REFERANSE,
                samvaersfradragDatoFraTil = Periode(datoFom = LocalDate.parse("2019-01-01"), datoTil = LocalDate.parse("2020-10-01")),
                barnPersonId = 1,
                barnFodselsdato = LocalDate.parse("2016-03-17"),
                samvaersklasse = "02",
            ),
        )
        samvaersfradragGrunnlagPeriodeListe.add(
            SamvaersfradragGrunnlagPeriode(
                referanse = TestUtil.SAMVÆRSFRADRAG_REFERANSE,
                samvaersfradragDatoFraTil = Periode(datoFom = LocalDate.parse("2019-01-01"), datoTil = LocalDate.parse("2020-10-01")),
                barnPersonId = 2,
                barnFodselsdato = LocalDate.parse("2017-05-17"),
                samvaersklasse = "02",
            ),
        )

        // Lag sjabloner
        val sjablonPeriodeListe = listOf(
            SjablonPeriode(
                Periode(datoFom = LocalDate.parse("2018-07-01"), datoTil = LocalDate.parse("2020-06-30")),
                Sjablon(
                    navn = SjablonNavn.SAMVÆRSFRADRAG.navn,
                    nokkelListe = listOf(
                        SjablonNokkel(navn = SjablonNøkkelNavn.SAMVÆRSKLASSE.navn, verdi = "02"),
                        SjablonNokkel(navn = SjablonNøkkelNavn.ALDER_TOM.navn, verdi = "5"),
                    ),
                    innholdListe = listOf(
                        SjablonInnhold(navn = SjablonInnholdNavn.ANTALL_DAGER_TOM.navn, verdi = BigDecimal.ZERO),
                        SjablonInnhold(navn = SjablonInnholdNavn.ANTALL_NETTER_TOM.navn, verdi = BigDecimal.valueOf(8)),
                        SjablonInnhold(navn = SjablonInnholdNavn.FRADRAG_BELØP.navn, verdi = BigDecimal.valueOf(727)),
                    ),
                ),
            ),
        )

        val beregnSamvaersfradragGrunnlag = BeregnSamvaersfradragGrunnlag(
            beregnDatoFra = LocalDate.parse("2020-06-01"),
            beregnDatoTil = LocalDate.parse("2020-07-01"),
            samvaersfradragGrunnlagPeriodeListe = samvaersfradragGrunnlagPeriodeListe,
            sjablonPeriodeListe = sjablonPeriodeListe,
        )

        val resultat = samvaersfradragPeriode.beregnPerioder(beregnSamvaersfradragGrunnlag)

        assertAll(
            Executable { assertThat(resultat.resultatPeriodeListe).hasSize(1) },
            Executable { assertThat(resultat.resultatPeriodeListe[0].resultatBeregningListe).hasSize(2) },
            Executable { assertThat(resultat.resultatPeriodeListe[0].resultatDatoFraTil.datoFom).isEqualTo(LocalDate.parse("2020-06-01")) },
            Executable { assertThat(resultat.resultatPeriodeListe[0].resultatDatoFraTil.datoTil).isEqualTo(LocalDate.parse("2020-07-01")) },
            Executable {
                assertThat(resultat.resultatPeriodeListe[0].resultatBeregningListe[0].belop).isEqualByComparingTo(
                    BigDecimal.valueOf(727),
                )
            },
            Executable {
                assertThat(resultat.resultatPeriodeListe[0].resultatBeregningListe[1].belop).isEqualByComparingTo(
                    BigDecimal.valueOf(727),
                )
            },
        )
    }

    @Test
    @DisplayName("Test med feil i grunnlag som skal resultere i avvik")
    fun testGrunnlagMedAvvik() {
        // Lag samværsinfo
        val samvaersklassePeriodeListe = listOf(
            SamvaersfradragGrunnlagPeriode(
                referanse = TestUtil.SAMVÆRSFRADRAG_REFERANSE,
                samvaersfradragDatoFraTil = Periode(datoFom = LocalDate.parse("2019-01-01"), datoTil = LocalDate.parse("2020-07-01")),
                barnPersonId = 1,
                barnFodselsdato = LocalDate.parse("2014-03-17"),
                samvaersklasse = "02",
            ),
        )

        // Lag sjabloner
        val sjablonPeriodeListe = mutableListOf<SjablonPeriode>()
        sjablonPeriodeListe.add(
            SjablonPeriode(
                sjablonPeriode = Periode(datoFom = LocalDate.parse("2018-07-01"), datoTil = LocalDate.parse("2020-06-30")),
                sjablon = Sjablon(
                    navn = SjablonNavn.SAMVÆRSFRADRAG.navn,
                    nokkelListe = listOf(
                        SjablonNokkel(navn = SjablonNøkkelNavn.SAMVÆRSKLASSE.navn, verdi = "02"),
                        SjablonNokkel(navn = SjablonNøkkelNavn.ALDER_TOM.navn, verdi = "5"),
                    ),
                    innholdListe = listOf(
                        SjablonInnhold(navn = SjablonInnholdNavn.ANTALL_DAGER_TOM.navn, verdi = BigDecimal.ZERO),
                        SjablonInnhold(navn = SjablonInnholdNavn.ANTALL_NETTER_TOM.navn, verdi = BigDecimal.valueOf(8)),
                        SjablonInnhold(navn = SjablonInnholdNavn.FRADRAG_BELØP.navn, verdi = BigDecimal.valueOf(727)),
                    ),
                ),
            ),
        )
        sjablonPeriodeListe.add(
            SjablonPeriode(
                Periode(LocalDate.parse("2018-07-01"), null),
                Sjablon(
                    navn = SjablonNavn.SAMVÆRSFRADRAG.navn,
                    nokkelListe = listOf(
                        SjablonNokkel(navn = SjablonNøkkelNavn.SAMVÆRSKLASSE.navn, verdi = "02"),
                        SjablonNokkel(navn = SjablonNøkkelNavn.ALDER_TOM.navn, verdi = "10"),
                    ),
                    innholdListe = listOf(
                        SjablonInnhold(navn = SjablonInnholdNavn.ANTALL_DAGER_TOM.navn, verdi = BigDecimal.ZERO),
                        SjablonInnhold(navn = SjablonInnholdNavn.ANTALL_NETTER_TOM.navn, verdi = BigDecimal.valueOf(8)),
                        SjablonInnhold(navn = SjablonInnholdNavn.FRADRAG_BELØP.navn, verdi = BigDecimal.valueOf(1052)),
                    ),
                ),
            ),
        )

        val beregnSamvaersfradragGrunnlag = BeregnSamvaersfradragGrunnlag(
            beregnDatoFra = LocalDate.parse("2018-07-01"),
            beregnDatoTil = LocalDate.parse("2021-01-01"),
            samvaersfradragGrunnlagPeriodeListe = samvaersklassePeriodeListe,
            sjablonPeriodeListe = sjablonPeriodeListe,
        )

        val avvikListe = samvaersfradragPeriode.validerInput(beregnSamvaersfradragGrunnlag)

        assertAll(
            Executable { assertThat(avvikListe).isNotEmpty() },
            Executable { assertThat(avvikListe).hasSize(2) },
            Executable {
                assertThat(
                    avvikListe[0].avvikTekst,
                ).isEqualTo("Første dato i samvaersfradragGrunnlagPeriodeListe (2019-01-01) er etter beregnDatoFom (2018-07-01)")
            },
            Executable { assertThat(avvikListe[0].avvikType).isEqualTo(Avvikstype.PERIODE_MANGLER_DATA) },
            Executable {
                assertThat(
                    avvikListe[1].avvikTekst,
                ).isEqualTo("Siste dato i samvaersfradragGrunnlagPeriodeListe (2020-07-01) er før beregnDatoTil (2021-01-01)")
            },
            Executable { assertThat(avvikListe[1].avvikType).isEqualTo(Avvikstype.PERIODE_MANGLER_DATA) },
        )
    }
}
